flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkBlueGreenDeploymentController.java

```diff
@@ -0,0 +1,703 @@
+            LOG.info("Using checkpoint: " + location);
+            adjustedSpec.getTemplate().getSpec().getJob().setInitialSavepointPath(location);
+        }
+
+        flinkDeployment.setSpec(adjustedSpec.getTemplate().getSpec());
+
+        // Deployment metadata
+        ObjectMeta flinkDeploymentMeta = getDependentObjectMeta(bgDeployment);
+        flinkDeploymentMeta.setName(childDeploymentName);
+        flinkDeploymentMeta.setLabels(
+                Map.of(deploymentType.getClass().getSimpleName(), deploymentType.toString()));
+        flinkDeployment.setMetadata(flinkDeploymentMeta);
+
+        // Deploy
+        josdkContext.getClient().resource(flinkDeployment).createOrReplace();
+    }
+
+    private static void deleteDeployment(
+            FlinkDeployment currentDeployment, Context<FlinkBlueGreenDeployment> josdkContext) {
+        String deploymentName = currentDeployment.getMetadata().getName();
+        List<StatusDetails> deletedStatus =
+                josdkContext
+                        .getClient()
+                        .resources(FlinkDeployment.class)
+                        .inNamespace(currentDeployment.getMetadata().getNamespace())
+                        .withName(deploymentName)
+                        .delete();
+
+        boolean deleted =
+                deletedStatus.size() == 1
+                        && deletedStatus.get(0).getKind().equals("FlinkDeployment");
+
+        if (!deleted) {
+            LOG.info("Deployment '" + deploymentName + "' not deleted, will retry");
+        } else {
+            LOG.info("Deployment '" + deploymentName + "' deleted!");
+        }
+    }
+
+    private ObjectMeta getDependentObjectMeta(FlinkBlueGreenDeployment bgDeployment) {
+        ObjectMeta bgMeta = bgDeployment.getMetadata();
+        ObjectMeta objectMeta = new ObjectMeta();
+        objectMeta.setNamespace(bgMeta.getNamespace());
+        objectMeta.setOwnerReferences(
+                List.of(
+                        new OwnerReference(
+                                bgDeployment.getApiVersion(),
+                                true,
+                                false,
+                                bgDeployment.getKind(),
+                                bgMeta.getName(),
+                                bgMeta.getUid())));
+        return objectMeta;
+    }
+
+    private static <T> T adjustNameReferences(
+            T spec,
+            String deploymentName,
+            String childDeploymentName,
+            String wrapperKey,
+            Class<T> valueType)
+            throws JsonProcessingException {
+        String serializedSpec = SpecUtils.serializeObject(spec, wrapperKey);
+        String replacedSerializedSpec = serializedSpec.replace(deploymentName, childDeploymentName);
+        return SpecUtils.deserializeObject(replacedSerializedSpec, wrapperKey, valueType);
+    }
+
+    public static void logAndThrow(String message) {
+        LOG.error(message);
```

Comment:
Do we need to log this here? Maybe better to simply throw it and let error handling logic log it if needed